sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -247,6 +253,10 @@ object ShuffleExchangeExec {
         val projection = UnsafeProjection.create(h.partitionIdExpression :: Nil, outputAttributes)
         row => projection(row).getInt(0)
       case RangePartitioning(_, _) | SinglePartition => identity
+      case LocalPartitioning(_, _) =>
+        (row: InternalRow) => {
+          TaskContext.get().partitionId()
```

Comment:
Can we try to do this once per partition instead of for each row?

Reply:
I think so. Will update in next commit.

Reply:
Fixed.